eslint-local-rules.js

```diff
@@ -33,4 +33,67 @@ module.exports = {
       }
     },
   },
+  'no-async-module-init': {
+    meta: {
+      type: 'problem',
+      docs: {
+        description: 'disallow async module initialisation',
+        category: 'Possible Errors',
+        recommended: true,
+      },
+      messages: {
+        noAsyncRegister:
+          'Disallowing static async {{ name }} function in modules to prevent unexpected startup failures or timeouts.',
+        noAsyncProviderFactory:
+          'Disallowing async useFactory in module providers to prevent unexpected startup failures or timeouts.',
+        noReturnPromiseDynamicModule:
+          'Disallowing static async functions with returnType Promise<DynamicModule> in modules to prevent unexpected startup failures or timeouts.',
+      },
+    },
+    schema: [],
+    create: function (context) {
+      const nameSymbols = [
+        'register',
+        'forRoot',
+        'forRootAsync',
+        'forFeature',
+        'forFeatureAsync',
+      ]
+
+      const rules = {
+        ...nameSymbols.reduce((rules, name) => {
+          return {
+            ...rules,
+            [`MethodDefinition[static=true][key.name='${name}'][value.async=true]`]: (
+              node,
+            ) => {
+              context.report({
+                node,
+                messageId: 'noAsyncRegister',
+                data: {
+                  name,
+                },
+              })
+            },
+          }
+        }, {}),
+        "MethodDefinition[static=true][value.async=true] TSTypeReference[typeName.name='Promise'] TSTypeReference[typeName.name='DynamicModule']": (
```

Comment:
Can skip the async check here.

```suggestion
        "MethodDefinition[static=true] TSTypeReference[typeName.name='Promise'] TSTypeReference[typeName.name='DynamicModule']": (
```